### PR TITLE
Fix flaky MessageIdTest and introduce some testing improvements

### PR DESCRIPTION
**Pull request by lhotari**
### Motivation

MessageIdTest class contains 2 of the flaky test cases in the Pulsar code base:
* `org.apache.pulsar.client.impl.MessageIdTest.testChecksumVersionComptability` and 
* `org.apache.pulsar.client.impl.MessageIdTest.testChecksumReconnection`

These test cases don't have much to do with `MessageId`, but are tests for validating message checksum handling in cases where there are pre 1.15 version brokers and post 1.15 version brokers in a mixed broker environment. The tests might not be very relevant any more. However it was taken as a learning experiment to fix these tests and refactor them so that the flakiness of the test code would be eliminated. Similar patterns might be needed in other tests to eliminate flakiness.

### Modifications

The changes aren't only to fix MessageIdTest. Most changes could help reduce flakiness of other tests as well.

Improve shutdown of the broker and related services to reduce test flakiness
- await for termination of executors
- close the listen channel synchronously
- use shutdown instead of shutdownNow in AbstractMetadataStore.close
  so that in-flight tasks get processed

Handle special case where the executor rejects the task and the callback was never called

Improve logging in MockedPulsarServiceBaseTest related to stopping and starting

Refactor PulsarClient initialization and lifecycle management in tests

Add getter and setter to access remoteEndpointProtocolVersion field
- it makes it easier to override for tests

Add hooks for overriding the producer implementation in PulsarClientImpl
- useful for tests. Instead of relying on Mockito, there's a pure Java
  way to inject behavior to producer implementations for testing purposes

Introduce PulsarTestClient that contains ways to prevent race conditions and test flakiness
- provides features for simulating failure conditions, for example
  the case of the broker connection disconnecting

Add solution for using Enums classes as source for TestNG DataProvider

Fix flaky MessageIdTest and move checksum related tests to new class

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java [hunk 1]
```diff
@@ -408,7 +408,7 @@ public void initializeFailed(ManagedLedgerException e) {
                         pendingLedger.ledger.asyncClose(new CloseCallback() {
                             @Override
                             public void closeComplete(Object ctx) {
-                                // no-op
+                                log.info("[{}] Closed pending initializer managed ledger.", name);
```
Comment (eolivelli) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:411]: probably we want to leave this at `debug` level

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java [hunk 2]
```diff
@@ -512,6 +512,7 @@ public void shutdown() throws InterruptedException, ManagedLedgerException {
             ledger.asyncClose(new AsyncCallbacks.CloseCallback() {
                 @Override
                 public void closeComplete(Object ctx) {
+                    log.info("[{}] Closed managed ledger.", ledger.getName());
```
Comment (eolivelli) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:515]: probably we want to leave this at `debug` level

File: pulsar-io/hbase/pom.xml [hunk 3]
```diff
@@ -70,12 +70,6 @@
             <version>${hbase.version}</version>
         </dependency>
 
-        <dependency>
```
Comment (eolivelli) [inline pulsar-io/hbase/pom.xml:73]: this is patch of another patch

**Review by eolivelli — COMMENTED** [review]
great work !

**Comment by eolivelli** [comment]
@codelipenghui  @zymap @sijie @rdhabalia this patch is going to fix the most Flaky test in the suite.

please take a look

thank you @lhotari  for contributing this work

File: pulsar-io/hbase/pom.xml (hunk 3 above)
Comment (lhotari) [reply pulsar-io/hbase/pom.xml:73]: yes I cherry-picked it since the fix for the broken master branch wasn't merged yet when I pushed this PR. I'll rebase (the commit would get remove) this PR.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java (hunk 1 above)
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:411]: good catch, I'll fix it.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java (hunk 2 above)
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:515]: good catch, I'll fix it.

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java [hunk 4]
```diff
@@ -207,7 +207,7 @@ public void accept(Notification n) {
 
     @Override
     public void close() throws Exception {
-        executor.shutdownNow();
+        executor.shutdown();
```
Comment (merlimat) [inline pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java:210]: Why do we need this change? The `shutdownNow()` will ensure that the thread is interrupted if it was waiting on anything. I think that, as a general rule, it should be preferable to wait for the thread to exit on its own.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java [hunk 5]
```diff
@@ -408,7 +408,7 @@ public void initializeFailed(ManagedLedgerException e) {
                         pendingLedger.ledger.asyncClose(new CloseCallback() {
                             @Override
                             public void closeComplete(Object ctx) {
-                                // no-op
+                                log.debug("[{}] Closed pending initializer managed ledger.", name);
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:411]: `if (log.isDebugEnabled()) { ...} `

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java [hunk 6]
```diff
@@ -512,6 +512,7 @@ public void shutdown() throws InterruptedException, ManagedLedgerException {
             ledger.asyncClose(new AsyncCallbacks.CloseCallback() {
                 @Override
                 public void closeComplete(Object ctx) {
+                    log.debug("[{}] Closed managed ledger.", ledger.getName());
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:515]: Same as above

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java [hunk 7]
```diff
@@ -542,8 +543,17 @@ public void closeFailed(ManagedLedgerException exception, Object ctx) {
         }
 
         scheduledExecutor.shutdownNow();
-        orderedExecutor.shutdownNow();
+        if (!scheduledExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
+           log.error("Cannot stop scheduledExecutor gracefully.");
+        }
+        orderedExecutor.shutdown();
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java:549]: Why not: 
```suggestion
        orderedExecutor.shutdownNow();
```

**Review by merlimat — COMMENTED** [review]
Very nice work, just a couple of comments on executors

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 8]
```diff
@@ -669,11 +669,23 @@ public void close() throws IOException {
         });
 
         if (listenChannel != null) {
-            listenChannel.close();
+            try {
+                if (!listenChannel.close().await(30, TimeUnit.SECONDS)) {
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:673]: Please make timeout configurable.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChecksumTest.java [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+import org.apache.pulsar.common.protocol.Commands.ChecksumType;
+import org.apache.pulsar.tests.EnumValuesDataProvider;
+import org.awaitility.Awaitility;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+public class MessageChecksumTest extends BrokerTestBase {
```
Comment (sijie) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChecksumTest.java:53]: This should be a separate pull request. Don't couple multiple changes together here.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PulsarTestClient.java [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+ * features:
+ * - can override remote endpoint protocol version in a thread safe manner
+ * - can reject new connections from the client to the broker
+ * - can drop all OpSend messages after they have been added to pendingMessages and processed
+ *   by the client. This simulates a situation where sending messages go to a "black hole".
+ * - can synchronize operations with the help of the pending message callback which gets
+ *   called after the message to send out has been added to the pending messages in the client.
+ *
+ */
+public class PulsarTestClient extends PulsarClientImpl {
```
Comment (sijie) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PulsarTestClient.java:54]: same comment above

**Review by sijie — CHANGES_REQUESTED** [review]

**Review by zymap — COMMENTED** [review]
nice work!

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java (hunk 4 above)
Comment (lhotari) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java:210]: Thanks for the review @merlimat .

> Why do we need this change?

Valid point. It's not actually needed. 

When I was debugging the flakiness, the key issue of the flakiness of MessageIdTest was related to the restarting of the broker not happening "cleanly". I started making an experiment where I'd revisit the shutdown of the broker. However these changes weren't relevant for making the test stable. I'll remove them completely out of this PR.

The relevant problem in shutdown was in BrokerService where `listenChannel.close()` doesn't use `.await` to synchronously wait for the completion of shutting down the listeners. I'll submit that fix as a separate PR. @sijie suggested in his review comments to split some changes as separate PRs.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 8 above)
Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:673]: Thanks for the review @sijie .
I'll move this change to a separate PR as you have suggested for one of the other changes.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChecksumTest.java (hunk 9 above)
Comment (lhotari) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageChecksumTest.java:53]: Thanks for the suggestion of splitting the changes to separate pull requests. I'll remove some changes from this PR to fix the flaky MessageIdTest. I was thinking that this change to split MessageChecksumTest from MessageIdTest is the relevant for fixing the flaky MessageIdTest, so I was thinking to to keep it in this PR, but move the other unrelated parts to new PRs. I hope this will be an acceptable way to resolve this.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PulsarTestClient.java (hunk 10 above)
Comment (lhotari) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/PulsarTestClient.java:54]: +1, I hope this gets addressed when I revisit the PR and remove unrelated changes. Currently the PulsarTestClient is needed to fix the flaky tests, so perhaps it could be kept in this PR after the unrelated changes have been removed?

File: pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java (hunk 4 above)
Comment (lhotari) [reply pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java]: I have now removed the executor shutdown/shutdownNow related changes from this PR. PTAL @merlimat

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java (hunk 6 above)
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java]: I have removed this change from this PR.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java (hunk 7 above)
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java]: I have removed this change from this PR.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 8 above)
Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java]: I have removed this change from this PR. The new PR for closing the channels is #9308 . PTAL @sijie

**Review by merlimat — APPROVED** [review]

**Comment by merlimat** [comment]
@lhotari It seems there's a genuine test failure: 

```
Error:  Failures: 
Error:  org.apache.pulsar.client.impl.PartitionedProducerImplTest.testCustomMessageRouterInstance(org.apache.pulsar.client.impl.PartitionedProducerImplTest)
[INFO]   Run 1: PASS
Error:    Run 2: PartitionedProducerImplTest.testCustomMessageRouterInstance:99->getMessageRouter:105 ? NullPointer
[INFO] 
Error:  org.apache.pulsar.client.impl.PartitionedProducerImplTest.testRoundRobinPartitionMessageRouterImplInstance(org.apache.pulsar.client.impl.PartitionedProducerImplTest)
[INFO]   Run 1: PASS
Error:    Run 2: PartitionedProducerImplTest.testRoundRobinPartitionMessageRouterImplInstance:89->getMessageRouter:105 ? NullPointer
[INFO] 
Error:  org.apache.pulsar.client.impl.PartitionedProducerImplTest.testSinglePartitionMessageRouterImplInstance(org.apache.pulsar.client.impl.PartitionedProducerImplTest)
[INFO]   Run 1: PASS
Error:    Run 2: PartitionedProducerImplTest.testSinglePartitionMessageRouterImplInstance:80->getMessageRouter:105 ? NullPointer
[INFO] 
```

**Comment by lhotari** [comment]
> It seems there's a genuine test failure: 

@merlimat Thanks for the heads up. I'll address it tomorrow.

**Review by sijie — APPROVED** [review]

**Review by eolivelli — APPROVED** [review]
LGTM
